src/common/libflux/attr.c

```diff
@@ -184,6 +189,34 @@ int flux_get_rank (flux_t *h, uint32_t *rank)
     return 0;
 }
 
+const char *flux_get_hostbyrank (flux_t *h, uint32_t rank)
+{
+    struct attr_cache *c;
+    const char *result;
+
+    if (rank == FLUX_NODEID_ANY)
+        return "any";
+    if (rank == FLUX_NODEID_UPSTREAM)
+        return "upstream";
+    if (!(c = get_attr_cache (h)))
+        return "memory-alloc-error"; // unlikely
+    if (!c->hostlist) {
+        const char *val;
+        if (!(val = flux_attr_get (h, "hostlist"))
+            || !(c->hostlist = hostlist_decode (val)))
+            goto error;
+    }
+    if (!(result = hostlist_nth (c->hostlist, rank)))
+        goto error;
+    return result;
+error:
+    snprintf (c->hostbuf,
```

Comment:
Wow, it is kind of a bummer the attr cache has to contain 64 bytes `hostbuf` just to print `rank=N` when `hostlist_nth()` fails. Perhaps not a big deal, but what if `flux_get_hostbyrank()` returns NULL on error with errno set? Most of the existing use cases _also_ print rank, and I think `%s` in a format will usually print `(null)` if the string argument is NULL...

Eh, not a big deal in the end, I just thought I'd bring it up.

Reply:
Now that I look at that, it does seem like overkill!

In case this gets used in other contexts besides printf, maybe returning "(null)" would be OK.